File: solid/l.cpp
#include <iostream>
#include <cstdint>

using namespace std;

/*
https://www.linkedin.com/pulse/first-5-principles-oop-c-what-so-solid-abdalrahman-hesham

Liskov Substitution Principle

This principle states that every derived class should be substitutable for their base
class. In other words, derived classes should behave correctly, as expected and avoid
logical errors due to their parent class.

Say we have a Rectangle class, which is a basic class that has width and height
attributes, set and get width functions, set and get height functions and a calculate area
function, as seen below.

what's the problem with the code below? Isn't the square correctly handling the dimension
inequality case?

Well, the problem is that the user using "SetWidth" and "SetHeight" functions might not
get that they change the other attribute, so the user might get unexpected results when
calling "CalculateArea" function.

So, to sum it up, This Square class isn't substitutable for the Rectangle class, and
thereby breaks Liskov principle. See c++ code below, then go back to the C# solution.

For example, if we were to write a test for the classes below the results would differ.
In C#: 
[TestMethod]
public void TwentyFourfor4x6RectanglefromSquare()
{
	Rectangle newRectangle = new Square();
	newRectangle.Height = 4;
	newRectangle.Width = 6;
	var result = AreaCalculator.CalculateArea(newRectangle);
	Assert.AreEqual(24, result);
}
 
The above test would fail because the expected result is 24, however, the actual area
calculated would be 36.
This is the problem. Even though the Square class is a subset of the Rectangle class, the
Object of Rectangle class is not substitutable by the object of the Square class without
causing a problem in the system. If the system adhered to the Liskov Substitution
Principle, you may avoid the above problem.

We can fix the issues like this:
public abstract class Shape
{
	public abstract int Area();
}
 
Here we have moved an abstract method Area in the Shape class and each subclass will give
its own definition to the Area method. Rectangle and Square class can be modified as shown
in the listing below:
public class Rectangle : Shape
{
	public int Height { get; set; }
	public int Width { get; set; }
	public override int Area()
	{
		return Height * Width;
	}
}
 
public class Square : Shape
{
	public int Sides;
	public override int Area()
	{
		return Sides * Sides;
	}
}

Then...
[TestMethod]
public void TwentyFourfor4x6Rectangleand9for3x3Square()
{
	var shapes = new List<Shape>{
		new Rectangle{Height=4,Width=6},
		new Square{Sides=3}
	};
	var areas = new List<int>();

	foreach (Shape shape in shapes)
	{
		areas.Add(shape.Area());
	}
	Assert.AreEqual(24, areas[0]);
	Assert.AreEqual(9, areas[1]);
}

C# example from:
https://www.infragistics.com/community/blogs/b/dhananjay_kumar/posts/simplifying-the-liskov-substitution-principle-of-solid-in-c

ALSO:
From: https://www.youtube.com/watch?v=Ntraj80qN2k&t=582s

Behavioral subtyping (aka “IS-A” relationship)
• Contravariance of method arguments in a subtype.
• Covariance of return types in a subtype.
• Preconditions cannot be strengthened in a subtype.
• Postconditions cannot be weakened in a subtype.
• Invariants of the super type must be preserved in a subtype.

Guideline: Make sure that inheritance is about behavior, not about data.
Guideline: Make sure that the contract of base types is adhered to.
Guideline: Make sure to adhere to the required concept.
*/

class Rectangle{
	protected:
		uint16_t Width;
		uint16_t Height;

	public:
		Rectangle(uint16_t width, uint16_t height): Width(width), Height(height){}

		virtual void SetWidth(uint16_t width){
			Width = width;
		}

		virtual void SetHeight(uint16_t height){
			Height = height;
		}

		virtual uint16_t GetWidth(){return Width;}
		virtual uint16_t GetHeight(){return Height;}
		virtual uint32_t CalculateArea(){return Width*Height;}
		virtual ~Rectangle() = default;
};

class Square : public Rectangle{
	public:
		Square(uint16_t L): Rectangle(L, L){}
		void SetWidth(uint16_t width) override{
			Width = width; Height = width;
		}
		void SetHeight(uint16_t height) override{
			Width = height; Height = height;
		}
		uint16_t GetWidth() override{return Width;}
		uint16_t GetHeight() override{return Height;}
};

File: solid/python/l.py
"""
https://www.pythontutorial.net/python-oop/python-liskov-substitution-principle/
The Liskov substitution principle states that a child class must be substitutable for its parent class. Liskov substitution principle aims to ensure that the child class can assume the place of its parent class without causing any errors.

Consider the following example:

from abc import ABC, abstractmethod


class Notification(ABC):
    @abstractmethod
    def notify(self, message, email):
        pass


class Email(Notification):
    def notify(self, message, email):
        print(f'Send {message} to {email}')


class SMS(Notification):
    def notify(self, message, phone):
        print(f'Send {message} to {phone}')


if __name__ == '__main__':
    notification = SMS()
    notification.notify('Hello', 'john@test.com')
    
In this example, we have three classes: Notification, Email, and SMS. The Email and SMS classes inherit from the Notification class.

The Notification abstract class has notify() method that sends a message to an email address.

The notify() method of the Email class sends a message to an email, which is fine.

However, the SMS class uses a phone number, not an email, for sending a message. Therefore, we need to change the signature of the notify() method of the SMS class to accept a phone number instead of an email.

The following NotificationManager class uses the Notification object to send a message to a Contact:

class Contact:
    def __init__(self, name, email, phone):
        self.name = name
        self.email = email
        self.phone = phone


class NotificationManager:
    def __init__(self, notification, contact):
        self.contact = contact
        self.notification = notification

    def send(self, message):
        if isinstance(self.notification, Email):
            self.notification.notify(message, contact.email)
        elif isinstance(self.notification, SMS):
            self.notification.notify(message, contact.phone)
        else:
            raise Exception('The notification is not supported')


if __name__ == '__main__':
    contact = Contact('John Doe', 'john@test.com', '(408)-888-9999')
    notification_manager = NotificationManager(SMS(), contact)
    notification_manager.send('Hello John')
    
The send() method of the NoticationManager class accepts a notification object. It checks whether the notification is an instance of the Email or SMS and passes the email and phone of contact to the notify() method respectively.
"""

# Conform with the Liskov substitution principle

from abc import ABC, abstractmethod

# First, redefine the notify() method of the Notification class so that it doesn’t include the email parameter:
class Notification(ABC):
    @abstractmethod
    def notify(self, message):
        pass

# Second, add the email parameter to the __init__ method of the Email class:
class Email(Notification):
    def __init__(self, email):
        self.email = email

    def notify(self, message):
        print(f'Send "{message}" to {self.email}')

# Third, add the phone parameter to the __init__ method of the SMS class:
class SMS(Notification):
    def __init__(self, phone):
        self.phone = phone

    def notify(self, message):
        print(f'Send "{message}" to {self.phone}')

class Contact:
    def __init__(self, name, email, phone):
        self.name = name
        self.email = email
        self.phone = phone

class NotificationManager:
    def __init__(self, notification):
        self.notification = notification

    def send(self, message):
        self.notification.notify(message)

if __name__ == '__main__':
    contact = Contact('John Doe', 'john@test.com', '(408)-888-9999')

    sms_notification = SMS(contact.phone)
    email_notification = Email(contact.email)

    notification_manager = NotificationManager(sms_notification)
    notification_manager.send('Hello John')

    notification_manager.notification = email_notification
    notification_manager.send('Hi John')